Composer/packages/client/src/store/persistence/FilePersistence.ts

```diff
@@ -1,165 +1,213 @@
-    const type = fileChangeType[action.type];
-    if (!type) return;
-    await this.operate({ ...type }, action.payload?.id, currentState);
+    return { id: `${file.id}${fileType}`, change: content, type: changeType };
   }
 
-  public registerHandleError(store: Store) {
-    const curStore = store;
-    this._handleError = name => err => {
-      //TODO: error handling now if sync file error, do a full refresh.
-      const fileName = name;
-      setError(curStore, {
-        message: err.response && err.response.data.message ? err.response.data.message : err,
-        summary: `HANDLE ${fileName} ERROR`,
+  private _getFileChangesById(
+    id: string,
+    changeType: ChangeType,
+    dialogs: DialogInfo[],
+    luFiles: LuFile[],
+    lgFiles: LgFile[]
+  ) {
+    const files: IFileChange[] = [];
+    const dialog = dialogs.find(dialog => dialog.id === id);
+    luFiles
+      .filter(lu => getBaseName(lu.id) === id)
+      .forEach(lu => {
+        files.push(this._createResourceInfo(lu, FileExtensions.Lu, changeType));
       });
-      fetchProject(curStore);
-    };
+    lgFiles
+      .filter(lg => getBaseName(lg.id) === id)
+      .forEach(lg => {
+        files.push(this._createResourceInfo(lg, FileExtensions.Lg, changeType));
+      });
+    files.push(this._createResourceInfo(dialog, FileExtensions.Dialog, changeType));
+    return files;
   }
 
-  // if create dialog, the lg and lu are created together
-  private async _doCreateForOtherFile(files: ResourceInfo[], extension: string, targetId: string) {
-    await Promise.all(
-      files
-        .filter(file => getBaseName(file.id) === targetId)
-        .map(async file => await this.doUpdate(`${file.id}${extension}`, file.content))
-    );
+  private _getFileChanges(changeInfo, id: string, previousState: State, currentState: State): IFileChange[] {
+    const { dialogs, luFiles, lgFiles } = currentState;
+    let files: IFileChange[] = [];
+    const { changeType, fileType } = changeInfo;
+    switch (fileType) {
+      case FileExtensions.Dialog: {
+        const dialog = dialogs.find(dialog => dialog.id === id);
+        if (changeType === ChangeType.CREATE) {
+          files = this._getFileChangesById(id, changeType, dialogs, luFiles, lgFiles);
+        } else if (changeType === ChangeType.DELETE) {
+          files = this._getFileChangesById(
+            id,
+            changeType,
+            previousState.dialogs,
+            previousState.luFiles,
+            previousState.lgFiles
+          );
+        } else {
+          files.push(this._createResourceInfo(dialog, fileType, changeType));
+        }
+        break;
+      }
+      case FileExtensions.Lu: {
+        const lu = luFiles.find(lu => lu.id === id);
+        files.push(this._createResourceInfo(lu, fileType, changeType));
+        break;
+      }
+      case FileExtensions.Lg: {
+        const lg = lgFiles.find(lg => lg.id === id);
+        files.push(this._createResourceInfo(lg, fileType, changeType));
```

Comment:
think about if you renaming this files to changes, this _createResourceInfo to createChange or something else, it would be simply reads as

changes.push(this.makeChange(...))

saying clearly you are construct a change object to a changes array.